src/openassetio-python/tests/bridge/python/test_converter.cpp

```diff
@@ -0,0 +1,268 @@
+  GIVEN("A python object") {
+    py::module_::import("openassetio");
+
+    // Create a python object and release it, to simulate an unmanaged
+    // PyObject being provided to us.
+    const py::object pyClass = py::module_::import("openassetio").attr("TraitsData");
+    py::object pyInstance = pyClass();
+    PyObject* pyTraitsData = pyInstance.release().ptr();
+    CHECK(Py_REFCNT(pyTraitsData) == 1);
+
+    WHEN("Python object is converted to a C++ object") {
+      TraitsDataPtr traitsData =
+          openassetio::python::converter::castFromPyObject<TraitsData>(pyTraitsData);
+
+      THEN("Python reference is obtained") { CHECK(Py_REFCNT(pyTraitsData) == 2); }
+
+      AND_WHEN("C++ reference is destroyed") {
+        traitsData.reset();
+
+        THEN("Python reference is released") { CHECK(Py_REFCNT(pyTraitsData) == 1); }
+      }
+    }
+
+    Py_DECREF(pyTraitsData);
+  }
+}
+
+SCENARIO("Attempting to cast to an unregistered type") {
+  GIVEN("an invalid for casting Python object") {
+    // Create a python object and release it, to simulate an unmanaged
+    // PyObject being provided to us.
+    const py::object pyClass = py::module_::import("decimal").attr("Decimal");
+    py::object pyInstance = pyClass(1.0);
+    // PyDecimal is not a type pybind has been registered to convert.
+    PyObject* pyDecimal = pyInstance.release().ptr();
+
+    /*
+     * Pybind error messages vary between release and debug mode:
+     * "Unable to cast Python instance of type <class 'decimal.Decimal'> to C++
+     * type 'openassetio::v1::hostApi::Manager'"
+     * vs.
+     * "Unable to cast Python instance to C++ type (compile in debug
+     * mode for details)"
+     */
+    WHEN("object is converted to a C++ object") {
+      REQUIRE_THROWS_WITH(
+          openassetio::python::converter::castFromPyObject<openassetio::hostApi::Manager>(
+              pyDecimal),
+          Catch::Matchers::StartsWith("Unable to cast Python instance"));
+    }
+
+    Py_DECREF(pyDecimal);
+  }
+}
+
+// NB: This test is excluded by default and is executed by explicitly
+// specifying the [no_openassetio_module] tag to the Catch2 runner
+// executable.
+// This is because we _must not_ have `_openassetio.so` loaded for these
+// tests.
+SCENARIO("Error attempting to convert API objects without openassetio module loaded",
+         "[.][no_openassetio_module]") {
+  GIVEN("A Python environment without openassetio loaded") {
+    REQUIRE_FALSE(py::module_::import("sys").attr("modules").contains("openassetio"));
+
+    AND_GIVEN("an OpenAssetIO C++ API object") {
+      const TraitsDataPtr traitsData = TraitsData::make();
+
+      WHEN("the C++ object is casted to a Python object") {
+        THEN("cast throws expected exception") {
+          REQUIRE_THROWS_WITH(openassetio::python::converter::castToPyObject(traitsData),
+                              std::string("Unregistered type : openassetio::v1::TraitsData"));
```

Comment:
[consider] And below, any way we can avoid baking `v1` in there to reduce future maintenance?

Reply:
Did a thing where the Stringified namespace macro is stripped from the expected string, I think it's pretty neat.

https://github.com/OpenAssetIO/OpenAssetIO/pull/929/commits/f2604f7d8ce1ca79fd2bfd5406e0f2425ae266b8

Reply:
Oh clever! I'll have to defer to @feltech as to the cunning involved :)

Reply:
Another alternative would be to build the expected value, rather than strip the actual value.  I.e.

`"Unregistered type : openassetio::" STRINGIFY(OPENASSETIO_CORE_ABI_VERSION) "::TraitsData"`

Reply:
Oh duh that's obviously better. b96ae62ea3cb732b190d559812cc8c88b67e45f6

(I did think about regex but std::regex has a deserved poor reputation so I try to avoid it.)

Reply:
LEBTM 👍  (@feltech feel free to close thread when happy)